scout/server/blueprints/institutes/controllers.py

```diff
@@ -514,3 +517,259 @@ def hgvs_str(gene_symbols, hgvs_p, hgvs_c):
+            term_title = None
+            if child in checkboxes:
+                custom_name = checkboxes[child].get("custom_name")
+                term_title = checkboxes[child].get("term_title")
+            term_obj = store.hpo_term(child)  # else it's an HPO term, and might have nested term:
+            node = None
+            try:
+                node = Node(child, parent=root, description=term_obj["description"])
+            except Exception as ex:
+                flash(f"Term {child} could not be find in database")
+                continue
+            all_terms[child] = term_obj
+            if custom_name:
+                node.custom_name = custom_name
+            if term_title:
+                node.term_title = term_title
+
+        # Rearrange tree nodes according the HPO ontology
+        root = store.organize_tree(all_terms, root)
+        LOG.info(f"Updated HPO tree:{root}:\n{RenderTree(root)}")
+        exporter = DictExporter()
+        for child_node in root.children:
+            # export node to dict
+            node_dict = exporter.export(child_node)
+            new_checkboxes[child_node.name] = node_dict
+
+    subpanel_obj["checkboxes"] = new_checkboxes
+    subpanel_obj["updated"] = datetime.datetime.now()
+    return subpanel_obj
+
+
+def phenomodel_checkgroups_filter(model_obj, user_form):
+    """Filter the checboxes of one or more subpanels according to preferences specified in the model preview
+
+    Args:
+        model_obj(dict): a dictionary coresponding to a phenotype model
+        user_form(request.form): a POST request form object
+
+    Returns:
+        model_obj(dict): an updated phenotype model dictionary to be saved to database
+    """
+    subpanels = model_obj.get("subpanels") or {}
+    check_list = (
+        user_form.getlist("cheked_terms") or []
+    )  # a list like this: ["subpanelID.rootTerm_checkedTerm", .. ]
+    updates_dict = {}
+    # From form values, create a dictionary like this: { "subpanel_id1":{"parent_term1":[children_terms], parent_term2:[children_terms2]}, .. }
+    for checked_value in check_list:
+        splitted_value = checked_value.split(".")
+        panel_key = splitted_value[0]
+        parent_term = splitted_value[1]
+        child_term = splitted_value[2]
+        if panel_key in updates_dict:
+            if parent_term in updates_dict[panel_key]:
+                updates_dict[panel_key][parent_term].append(child_term)
+            else:
+                updates_dict[panel_key][parent_term] = [child_term]
+        else:
+            updates_dict[panel_key] = {parent_term: [child_term]}
+
+    # loop over the subpanels of the model, and check they need to be updated
+    for key, subp in subpanels.items():
+        if key in updates_dict:  # if subpanel requires changes
+            model_obj["subpanels"][key] = _update_subpanel(subp, updates_dict[key])  # update it
+
+    return model_obj
+
+
+def _add_subpanel(model_id, model_obj, user_form):
```

Comment:
Here if not before the distinction between model_obj and model dicts become a bit blurred compared to how we usually do it. The obj should normally already have an id. You seem to be able to keep track anyway, so perhaps just consider renaming some model_obj arguments model or model_dict?

Reply:
Renaming all model_obj to model_dict

Reply:
Done